src/algorithm/BfsIterator.java

```diff
@@ -0,0 +1,70 @@
+package algorithm;
+
+import graph.Graph;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.NoSuchElementException;
+import java.util.Queue;
+import java.util.Set;
+
+public class BfsIterator implements Iterator<Integer> {
+
+  private final Graph graph;
+  private final int start;
+  private Set<Integer> visited;
+  private Queue<Integer> queue;
+  private int count;
+
+  public BfsIterator(Graph graph, int start) {
+    this.graph = graph;
+    this.start = start;
+    this.visited = new HashSet<>();
+    this.queue = new LinkedList<>();
+    this.count = 0;
+
+    if (!graph.contains(start)) {
+      throw new NoSuchElementException();
+    }
+    this.queue.add(start);
+    this.visited.add(start);
+  }
+
+  public Graph getGraph() {
+    return graph;
+  }
+
+  public int getStart() {
+    return start;
+  }
+
+  @Override
+  public Integer next() {
+    if (queue.isEmpty()) {
+      Collection<Integer> ids = this.graph.getVertexIDs();
+      ids.removeAll(this.visited);
```

Comment:
Why not use unvisited? Then you do not need to iterate the vertices set to remove the unvisited vertices.

Reply:
Also, I think you can do it smarter by making queue as a Collection. Then for BfsIterator you can construct it as a Queue, while for DfsIterator, you can make it a Stack. Then you can just use one  code for both. The following sample code is for your reference:

public class Test {
  
    public Collection<Integer> queue;
    
    public Test(boolean isBFS) {
      if (isBFS) {
        queue = new LinkedList<Integer>();
      } else {
        queue = new Stack<Integer>();
      }
    }
	
	public static void main(String[] args){
		Test test1 = new Test(true);
		Queue<Integer> q = (Queue<Integer>) test1.queue;
		q.offer(1);
		q.offer(2);
		q.poll();
		q.forEach(e -> {System.out.println(e);});
		
		Test test2 = new Test(false);
		Stack<Integer> s = (Stack<Integer>) test2.queue;
		s.push(1);
		s.push(2);
		s.pop();
		s.forEach(e-> {System.out.println(e);});
	}
}